[2.9 ACRA] crash adding bitmap to card

https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/47c1cab0-f57c-4162-aa09-f09d416bd11b

```
java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=2, result=-1, data=Intent { act=inline-data (has extras) }} to activity {com.ichi2.anki/com.ichi2.anki.multimediacard.activity.MultimediaEditFieldActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean android.graphics.Bitmap.compress(android.graphics.Bitmap$CompressFormat, int, java.io.OutputStream)' on a null object reference
at android.app.ActivityThread.deliverResults(ActivityThread.java:4382)
at android.app.ActivityThread.handleSendResult(ActivityThread.java:4426)
at android.app.ActivityThread.-wrap20(Unknown Source:0)
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1685)
at android.os.Handler.dispatchMessage(Handler.java:106)
at android.os.Looper.loop(Looper.java:164)
at android.app.ActivityThread.main(ActivityThread.java:6626)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:811)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean android.graphics.Bitmap.compress(android.graphics.Bitmap$CompressFormat, int, java.io.OutputStream)' on a null object reference
at com.ichi2.anki.multimediacard.fields.BasicImageFieldController.rotateAndCompress(BasicImageFieldController.java:6)
at com.ichi2.anki.multimediacard.fields.BasicImageFieldController.onActivityResult(BasicImageFieldController.java:9)
at com.ichi2.anki.multimediacard.activity.MultimediaEditFieldActivity.onActivityResult(MultimediaEditFieldActivity.java:3)
at android.app.Activity.dispatchActivityResult(Activity.java:7305)
at android.app.ActivityThread.deliverResults(ActivityThread.java:4378)
... 9 more
```

--- a/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java
@@ -28,19 +28,25 @@ import android.content.Intent;
 import android.content.pm.PackageManager;
 import android.database.Cursor;
 import android.graphics.Bitmap;
+import android.graphics.Color;
 import android.net.Uri;
 import android.os.Build;
 import android.provider.MediaStore;
 import android.provider.MediaStore.MediaColumns;
 import androidx.core.content.ContextCompat;
 import androidx.core.content.FileProvider;
+
+import android.text.format.Formatter;
 import android.util.DisplayMetrics;
+import android.view.Gravity;
 import android.view.View;
 import android.view.ViewGroup;
 import android.widget.Button;
+import android.widget.EditText;
 import android.widget.ImageView;
 import android.widget.LinearLayout;
 import android.widget.LinearLayout.LayoutParams;
+import android.widget.TextView;
 
 import com.ichi2.anki.R;
 import com.ichi2.compat.CompatHelper;
@@ -64,6 +70,8 @@ public class BasicImageFieldController extends FieldControllerBase implements IF
     private static final int IMAGE_SAVE_MAX_WIDTH = 1920;
 
     private ImageView mImagePreview;
+    private TextView mImageFileSize;
+    private TextView mImageFileSizeWarning;
 
     private String mTempCameraImagePath;
     private DisplayMetrics mMetrics = null;
@@ -97,6 +105,24 @@ public class BasicImageFieldController extends FieldControllerBase implements IF
 
         mImagePreview.setMaxHeight((int) Math.round(height * 0.4));
         mImagePreview.setMaxWidth((int) Math.round(width * 0.6));
+
+        mImageFileSize = new EditText(context);
+        mImageFileSize.setMaxWidth((int) Math.round(width * 0.6));
+        mImageFileSize.setEnabled(false);
+        mImageFileSize.setGravity(Gravity.CENTER_HORIZONTAL);
+        mImageFileSize.setBackground(null);
+        mImageFileSize.setVisibility(View.GONE);
+
+        //#5513 - Image compression failed, but we'll confuse most users if we tell them that. Instead, just imply that
+        //there's an action that they can take.
+        mImageFileSizeWarning = new EditText(context);
+        mImageFileSizeWarning.setMaxWidth((int) Math.round(width * 0.6));
+        mImageFileSizeWarning.setEnabled(false);
+        mImageFileSizeWarning.setTextColor(Color.parseColor("#FF4500")); //Orange-Red
+        mImageFileSizeWarning.setGravity(Gravity.CENTER_HORIZONTAL);
+        mImageFileSizeWarning.setVisibility(View.GONE);
+        mImageFileSizeWarning.setBackground(null);
+        mImageFileSizeWarning.setText(R.string.multimedia_editor_image_compression_failed);
 
         Button mBtnGallery = new Button(mActivity);
         mBtnGallery.setText(gtxt(R.string.multimedia_editor_image_field_editing_galery));
@@ -154,6 +180,8 @@ public class BasicImageFieldController extends FieldControllerBase implements IF
         }
 
         layout.addView(mImagePreview, ViewGroup.LayoutParams.MATCH_PARENT, p);
+        layout.addView(mImageFileSize, ViewGroup.LayoutParams.MATCH_PARENT);
+        layout.addView(mImageFileSizeWarning, ViewGroup.LayoutParams.MATCH_PARENT);
         layout.addView(mBtnGallery, ViewGroup.LayoutParams.MATCH_PARENT);
         layout.addView(mBtnCamera, ViewGroup.LayoutParams.MATCH_PARENT);
     }
@@ -179,6 +207,7 @@ public class BasicImageFieldController extends FieldControllerBase implements IF
         if (resultCode == Activity.RESULT_CANCELED) {
             return;
         }
+        mImageFileSizeWarning.setVisibility(View.GONE);
         if (requestCode == ACTIVITY_SELECT_IMAGE) {
             Uri selectedImage = data.getData();
             // Timber.d(selectedImage.toString());
@@ -220,6 +249,13 @@ public class BasicImageFieldController extends FieldControllerBase implements IF
         String outPath = inPath.substring(0, inPath.lastIndexOf(".")) + ".png";
         // Load into a bitmap with max size of 1920 pixels and rotate if necessary
         Bitmap b = BitmapUtil.decodeFile(f, IMAGE_SAVE_MAX_WIDTH);
+        if (b == null) {
+            //#5513 - if we can't decode a bitmap, return the original image
+            //And display a warning to push users to compress manually.
+            mImageFileSizeWarning.setVisibility(View.VISIBLE);
+            return inPath;
+        }
+
         FileOutputStream out = null;
         try {
             out = new FileOutputStream(outPath);
@@ -248,6 +284,8 @@ public class BasicImageFieldController extends FieldControllerBase implements IF
             Bitmap b = BitmapUtil.decodeFile(f, maxsize);
             b = ExifUtil.rotateFromCamera(f, b);
             mImagePreview.setImageBitmap(b);
+            mImageFileSize.setVisibility(View.VISIBLE);
+            mImageFileSize.setText(Formatter.formatFileSize(mActivity, f.length()));
         }
     }
 

--- a/AnkiDroid/src/main/java/com/ichi2/utils/BitmapUtil.java
+++ b/AnkiDroid/src/main/java/com/ichi2/utils/BitmapUtil.java
@@ -26,6 +26,8 @@ import android.graphics.drawable.Drawable;
 
 import android.widget.ImageView;
 
+import com.ichi2.anki.AnkiDroidApp;
+
 import java.io.File;
 import java.io.FileInputStream;
 import java.io.IOException;
@@ -71,7 +73,8 @@ public class BitmapUtil {
             }
 
         } catch (IOException e) {
-            // do nothing
+            //#5513 - We don't know the reason for the crash, let's find out.
+            AnkiDroidApp.sendExceptionReport(e, "BitmapUtil decodeFile");
         }
         return bmp;
     }

--- a/AnkiDroid/src/main/java/com/ichi2/utils/BitmapUtil.java
+++ b/AnkiDroid/src/main/java/com/ichi2/utils/BitmapUtil.java
@@ -41,9 +41,15 @@ public class BitmapUtil {
             BitmapFactory.Options o = new BitmapFactory.Options();
             o.inJustDecodeBounds = true;
 
-            FileInputStream fis = new FileInputStream(theFile);
-            BitmapFactory.decodeStream(fis, null, o);
-            fis.close();
+            FileInputStream fis = null;
+            try {
+                fis = new FileInputStream(theFile);
+                BitmapFactory.decodeStream(fis, null, o);
+            } finally {
+                if (fis != null) {
+                    fis.close();
+                }
+            }
 
             int scale = 1;
             if (o.outHeight > IMAGE_MAX_SIZE || o.outWidth > IMAGE_MAX_SIZE) {
@@ -56,10 +62,14 @@ public class BitmapUtil {
             // Decode with inSampleSize
             BitmapFactory.Options o2 = new BitmapFactory.Options();
             o2.inSampleSize = scale;
-            fis = new FileInputStream(theFile);
-            bmp = BitmapFactory.decodeStream(fis, null, o2);
 
-            fis.close();
+            try {
+                fis = new FileInputStream(theFile);
+                bmp = BitmapFactory.decodeStream(fis, null, o2);
+            } finally {
+                fis.close(); //don't need a null check, as we reuse the variable.
+            }
+
         } catch (IOException e) {
             // do nothing
         }

--- a/AnkiDroid/src/main/res/values/02-strings.xml
+++ b/AnkiDroid/src/main/res/values/02-strings.xml
@@ -241,4 +241,5 @@
     <string name="export_v2_forbidden">Please switch to the V1 scheduler before exporting a single deck with scheduling information.</string>
 
     <string name="activity_start_failed">The system does not have an app installed that can perform this action.</string>
+    <string name="multimedia_editor_image_compression_failed"><![CDATA[Camera images may be large. You may wish to compress & resize images in the media directory]]></string>
 </resources>

